### PR TITLE
Add annotate ops to sequence fuzz tests 

### PR DESCRIPTION
**Pull request by jzaffiro**
Before adding handles to the sequence fuzz tests, add annotateRange ops as a place to insert the handles. Because of updates to the equivalence function, I think some of the existing interval collection bugs are now showing up differently than before (sliding preference not equal --> property sets have different lengths).

### CODE REVIEW

File: packages/dds/sequence/src/test/intervalTestUtils.ts [hunk 1]
```diff
@@ -107,6 +108,25 @@ export function assertEquivalentSharedStrings(a: SharedString, b: SharedString)
 	}
 }
 
+function assertPropertiesEqual(a: SharedString, b: SharedString): void {
+	for (let i = 0; i < a.getLength(); i++) {
+		const aProps = a.getPropertiesAtPosition(i) ?? [];
+		const bProps = b.getPropertiesAtPosition(i) ?? [];
+		const aKeys = aProps === undefined ? [] : Object.keys(aProps);
```
Comment (Abe27342) [inline packages/dds/sequence/src/test/intervalTestUtils.ts:115]: I think just doing `assert.deepEqual(aProps, bProps)` will produce a better error message than going through them key-by-key. Simpler code too :)

File: packages/dds/test-dds-utils/src/ddsFuzzHarness.ts [hunk 2]
```diff
@@ -33,7 +33,7 @@ import {
 import { AttachState } from "@fluidframework/container-definitions";
 import type { IFluidHandle } from "@fluidframework/core-interfaces";
 import { unreachableCase } from "@fluidframework/core-utils/internal";
-import type { IChannelFactory, IChannelServices } from "@fluidframework/datastore-definitions";
+import type { IChannelFactory, IChannelServices } from "@fluidframework/datastore-definitions/internal";
```
Comment (Abe27342) [inline packages/dds/test-dds-utils/src/ddsFuzzHarness.ts:36]: nit: stray changes to this package and test-runtime-utils

**Review by Abe27342 — APPROVED** [review]
looks reasonable. In response to the description around different failure modes, note that you changed the weights of generating different options for the base fuzz test, so tests will be doing different things now.

Is there filed work for resolving the eventual consistency bugs here? If not, file some.

File: packages/dds/sequence/src/test/fuzz/fuzzUtils.ts [hunk 3]
```diff
@@ -303,6 +314,17 @@ export function createSharedStringGeneratorOperations(
 		};
 	}
 
+	async function annotateRange(state: ClientOpState): Promise<AnnotateRange> {
+		const { random } = state;
+		const key = random.pick(options.propertyNamePool);
+		const value = random.string(5);
```
Comment (anthony-murphy) [inline packages/dds/sequence/src/test/fuzz/fuzzUtils.ts:320]: we also need to support the value `undefined` to cover removing annotations

Comment (anthony-murphy) [reply packages/dds/sequence/src/test/fuzz/fuzzUtils.ts:320]: actually we should use both null and undefined as they have subtly different behavior

File: packages/dds/sequence/src/test/intervalTestUtils.ts (hunk 1 above)
Comment (jzaffiro) [reply packages/dds/sequence/src/test/intervalTestUtils.ts:115]: Doing this gives me a failure in the afterEach hook. I thought it had something to do with the undefined/null properties, but I get the same failure even without those as potential values. The same seeds fail with the key-by-key comparison as the deepEqual comparison, so my guess is that it's the same issue, just failing in a different place. I wasn't sure if I should still change the assert given the error message, though.

File: packages/dds/sequence/src/test/fuzz/intervalCollection.fuzz.spec.ts [hunk 4]
```diff
@@ -50,7 +50,7 @@ describe("IntervalCollection with stashing", () => {
 			stashableClientProbability: 0.2,
 		},
 		// AB#7220
-		skip: [22, 56, 86],
+		skip: [4, 11, 18, 29, 35, 41, 47, 53, 65, 66, 73, 76, 86, 93, 96, 99],
```
Comment (jzaffiro) [inline packages/dds/sequence/src/test/fuzz/intervalCollection.fuzz.spec.ts:53]: Both interval collection with stashing and shared string with stashing have a lot of failing seeds - does this point to a larger problem than the existing bugs we have open?

File: packages/dds/sequence/src/test/fuzz/fuzzUtils.ts [hunk 5]
```diff
@@ -303,6 +314,17 @@ export function createSharedStringGeneratorOperations(
 		};
 	}
 
+	async function annotateRange(state: ClientOpState): Promise<AnnotateRange> {
+		const { random } = state;
+		const key = random.pick(options.propertyNamePool);
+		const value = random.pick([random.string(5), typeof undefined, typeof null]);
```
Comment (anthony-murphy) [inline packages/dds/sequence/src/test/fuzz/fuzzUtils.ts:320]:
```suggestion
		const value = random.pick([random.string(5), undefined, null]);
```

Comment (anthony-murphy) [reply packages/dds/sequence/src/test/fuzz/fuzzUtils.ts:320]: using typeof will get the type of the value as a string, but we want the actual value

File: packages/dds/sequence/src/test/intervalTestUtils.ts [hunk 6]
```diff
@@ -107,6 +108,25 @@ export function assertEquivalentSharedStrings(a: SharedString, b: SharedString)
 	}
 }
 
+function assertPropertiesEqual(a: SharedString, b: SharedString): void {
+	for (let i = 0; i < a.getLength(); i++) {
+		const aProps = a.getPropertiesAtPosition(i) ?? [];
```
Comment (anthony-murphy) [inline packages/dds/sequence/src/test/intervalTestUtils.ts:113]:
```suggestion
		const aProps = a.getPropertiesAtPosition(i);
```

File: packages/dds/sequence/src/test/intervalTestUtils.ts [hunk 7]
```diff
@@ -107,6 +108,25 @@ export function assertEquivalentSharedStrings(a: SharedString, b: SharedString)
...
+		const bProps = b.getPropertiesAtPosition(i) ?? [];
```
Comment (anthony-murphy) [inline packages/dds/sequence/src/test/intervalTestUtils.ts:114]:
```suggestion
		const aProps = a.getPropertiesAtPosition(i);
		const bProps = b.getPropertiesAtPosition(i);
```

Comment (anthony-murphy) [reply packages/dds/sequence/src/test/intervalTestUtils.ts:114]: we shouldn't default the value here, as they should really be the same. if one is undefined, and the other is empty the check should fail

File: packages/dds/sequence/src/test/fuzz/intervalCollection.fuzz.spec.ts (hunk 4 above)
Comment (anthony-murphy) [reply packages/dds/sequence/src/test/fuzz/intervalCollection.fuzz.spec.ts:53]: possibly. i think i'm ok with the skips in this pr, as we are going to fix these issue shortly after correct?

File: packages/dds/sequence/src/test/intervalTestUtils.ts (hunk 1 above)
Comment (anthony-murphy) [reply packages/dds/sequence/src/test/intervalTestUtils.ts:115]: deep equal also won't work once we have handles, so we'll need to check key by key anyway

File: packages/dds/sequence/src/test/fuzz/intervalCollection.fuzz.spec.ts (hunk 4 above)
Comment (jzaffiro) [reply packages/dds/sequence/src/test/fuzz/intervalCollection.fuzz.spec.ts:53]: Correct, I'm planning to look into them once I finish this work

Comment (Abe27342) [reply packages/dds/sequence/src/test/fuzz/intervalCollection.fuzz.spec.ts:53]: I think that's fine. One thing I like to do as a sniff test when I'm unsure if the changes I've made are correct is minimizing / writing a minimal repro for one of the skipped seeds as part of the PR that skips it. This is also convenient for filing bugs (you can just point at the skipped but checked in test).

Your changes in this PR are pretty safe though. Assuming the failures you get have call stacks which actually suggest eventual consistency issues and don't set off alarm bells based on the changes here, this seems unnecessary for this change.

File: packages/dds/sequence/src/test/intervalTestUtils.ts (hunk 7 above)
Comment (anthony-murphy) [reply packages/dds/sequence/src/test/intervalTestUtils.ts:114]: if you need a default it should be to `{}` not `[]`

**Review by anthony-murphy — APPROVED** [review]

**Pull request merged**